### ISSUE TITLE
Cross platform compatible `.insert()` method.

### ISSUE THREAD

**Issue by elhigu** [tgriesser/knex#2682]
# Environment

Knex version: 0.15.x
Database + version: all 
OS: all

# Feature discussion / request

1. Explain what is your use case / 2. Explain what kind of feature would support this

This feature has been requested few times, but there hasn't been any proper feature request before this.

Libraries/apps that are using knex as a layer to write cross db compatible queries.

Currently different dialects doesn't have any crossdb compatible insert builder method, which returns `id` of created row as a return value. Currently for example postgresql requires that one adds `.returning('id')` to get that. 

If it is possible to be implemented to work also for multiple rows would be better, but it might not be possible for every DB without multiple queries. So probably would be enough if it work only with single row.

3. Give some API proposal, how the feature should work

Maybe .insert() could have second parameter (like it is currently used for giving returning columns) where for postgres and other databases which requires that one gives column name of id field one also defines the column name.

```js
.insert({ ... row ... }, { returningId: 'id' })
```

On databases where 'id' column is not necessary to be given tha name of the column is just ignored and default id is returned.

**Comment by ricardograca** [comment]
Multiple returning columns across all dialects would be great, even if it means extra queries in some of them.

**Comment by elhigu** [comment]
@ricardograca yeah, that would be nice, but it doesn't fit well to knex's current design principles, where builder always builds just a single query (except for migration builder, which can generate multiple).

Also implementing that probably would require that first query is executed before the second one, which is impossible without real DB connection. 

I would leave that part for ORMs. (which I think pretty much everyone should use on top of knex when building real applications)

**Comment by ricardograca** [comment]
That's a good point. Maybe another feature request would be to have `.returning()` always work across all dialects then, although that can also be abstracted away with an ORM, but it would simplify the work downstream.

**Comment by elhigu** [comment]
> Maybe another feature request would be to have .returning() always work across all dialects the

That would also require multiple queries, so I don't think that is possible either to do in knex. Another issue could be opened to discuss more about that though, since it is out of scope from this one.

**Comment by rijkvanzanten** [comment]
I personally think that @ricardograca's idea of making `.returning()` cross platform compatible solves for the same use case as this extension on the `.insert()` method. 

> Multiple returning columns across all dialects would be great, even if it means extra queries in some of them.

+1 to this, to me, `.returning()` should allow either a string or an array of strings (non-breaking-change): `.returning('id')`, `.returning(['id', 'name']);`

That being said, making `.returning()` cross-platform is a different feature request than supporting multiple columns in `.returning()` 🙂 

> but it doesn't fit well to knex's current design principles, where builder always builds just a single query (except for migration builder, which can generate multiple).

This is a great thing to strive for and aspire to, but I think it's hindering the usability of Knex more than it's worth.. Knex is by far the best sql abstraction layer for cross platform apps, but it has a handful of less-than-ideal quirks, like the differences in what it returns on insert.

To me, a big part of the decision to rely on Knex is that I don't end up doing something like

```
if (driver == 'mysql') {
  mysql.query('SELECT');
} else if (driver == 'pg') {
  pg.query();
} else if (driver == 'etc') { etc }
```

in my application logic. 

You could say I almost treat it as a jQuery for databases; abstracting away the quirks between the different database drivers so I can focus on my business logic. Being able to _reliably_ do:

```ts
const primaryKeys = await knex
  .insert([{ name: 'Rijk' }, { name: 'Mikael' }, { name: 'Ricardo' }])
  .into('people')
  .returning('id');
```

is a crucial part of that.

> Also implementing that probably would require that first query is executed before the second one, which is impossible without real DB connection.

Could you elaborate on this @elhigu?

> I would leave that part for ORMs

While I agree that ORMs and other libraries wrapping Knex can provide a lot of extra niceties to the database, I don't think we should draw the line for where Knex ends here. To me, Knex should provide a feature set that can reliably be used cross platform, and leave everything on top of that over to external libraries and packages. I don't think ensuring a reliable return type cross platform can be considered an extra on-top layer.

> (which I think pretty much everyone should use on top of knex when building real applications)

I think this is an over-generalization. Sure, when building an app where the schema is known and fixed, one could say it's advisable to use an ORM, but you can't simply say everyone should use one. For example, [in my current use case](https://github.com/directus/next), the database schema changes constantly in real time. Dynamically registering/unregistering models in something like Bookshelf is pretty cumbersome, where "raw Knex" combined with [the schema inspector](https://github.com/knex/knex-schema-inspector) offers just the right amount of flexibility in query building to dynamically request the right amount of data from the right sources.

> That would also require multiple queries, so I don't think that is possible either to do in knex

I believe this ties into the previous point "_[...] which is impossible without real DB connection."_. Is there a foundational design decision in Knex that prevents multiple queries to be run in succession @elhigu? 

---

The weird thing to me is the seemingly double standard around the level of abstraction that Knex provides. On the one hand, Knex abstracts away a lot of little things between the database drivers, like being able to insert `true` as a value in MySQL, but on the other hand, some features that would benefit greatly from some abstraction, like this returning discussion, are left as is.

In any case, keep up the great work y'all are doing with Knex ❤️

**Comment by elhigu** [comment]
> This is a great thing to strive for and aspire to, but I think it's hindering the usability of Knex more than it's worth.. Knex is by far the best sql abstraction layer for cross platform apps, but it has a handful of less-than-ideal quirks, like the differences in what it returns on insert. 

Complex,  non-trivial compatibility layers, should be done outside of knex as an external npm packages, where scope of those features and implementation will be well separated from the minimal core query building functionality of knex. It allows knex core functionality being made more robust without need to try to handle cornercase etc. issues of unnecessary higher level. Also in that way those packages can have more easily different authors / developers, who doesn't need to understand all knex internals.

>> Also implementing that probably would require that first query is executed before the second one, which is impossible without real DB connection.

> Could you elaborate on this @elhigu?

To get list of id's of inserted items, one query must be executed first before, the one which actually fetches those rows from db to emulate the `.returning(...)` functionality. If DB driver allows to pass multiple statements to DB server with still using parameter bindings, it can be done as a "single query", but that is not granted for all DB drivers.

>> I would leave that part for ORMs

> While I agree that ORMs and other libraries wrapping Knex can provide a lot of extra niceties to the database, I don't think we should draw the line for where Knex ends here. To me, Knex should provide a feature set that can reliably be used cross platform, and leave everything on top of that over to external libraries and packages. I don't think ensuring a reliable return type cross platform can be considered an extra on-top layer.

I think we should draw a line for where knex ends to it being query builder in a way that for its query builder instance it should return a single SQL + bindings that can be executed. Everything above that should be done in higher abstraction layers, that uses knex under the hood. It will allow keeping knex small, tidy and unopinionated and leave the complex stuff needing more code and logic to other packages. 

> Is there a foundational design decision in Knex that prevents multiple queries to be run in succession @elhigu?

The most important design decision behind that is that if we allow knex to create actually big multi statement SQL scripts for executing a single query builder it will open can of worms for implementing all complicated things making knex to be big chunk of bloat.

Another important design decision is that if something can be done outside of knex and is kind of opinionated functionality, it is better to implement as a separate package. For that we added extension support for knex. When some people wanted to support certain loggers, we added log hooks to allow people to use what ever they wanted. When snake case / camel case support was requested, we added hooks to allow modify identifier interpolation and post processing returned data, so that people can do their own features outside of knex.

Even with current really limited scope knex seems to be too big to be handled well with current contributors. With less bloat maybe we could even implement some important core features which would allow people implementing higher level packages to do better things (couple big things would be query introspection to allow checking what syntactic elements query builder has and some tree based presentation of the builder to allow doing more transformations for created query).

> The weird thing to me is the seemingly double standard around the level of abstraction that Knex provides. On the one hand, Knex abstracts away a lot of little things between the database drivers, like being able to insert true as a value in MySQL..

I believe that is MySQL driver feature, since knex doesn't seem to touch those input values https://runkit.com/embed/0rmailbw2l18 

Original thought of knex indeed was being a perfect emulation layer between databases, but later on it was apparent that even doing the low level query building well for 5 databases is enough hard and adding non trivial emulation is many times bigger task.

**Comment by rijkvanzanten** [comment]
Thanks for the in-depth insight @elhigu 🙂 This all makes perfect sense 👍🏻 

With all of this in mind, I can't help but wonder: should a next major version of Knex drop the dependency on the actual DB connection altogether in favor of emitting "clean" SQL queries that can then be executed using whatever third party package? That would alleviate a lot of overhead in the codebase, and prevents issues like https://github.com/knex/knex/issues/3963.

**Comment by kibertoad** [comment]
@rijkvanzanten Isn't that already possible by using .toSQL/.generateDdlCommands?

**Comment by rijkvanzanten** [comment]
@kibertoad Yeah! The use case is already possible, I was just wondering what the benefits are of having the actual connections configured through Knex itself if the main focus is generating SQL queries. It seems like working around/supporting different connection libraries adds quite a bit of overhead that could be seen as out-of-scope for a query builder. (Same thing for schema manipulation and migrations honestly!)

**Comment by kibertoad** [comment]
@rijkvanzanten Why would an average user want to manually pass generated SQL to a driver or use some additional wrapper on top of knex to execute the statement?
If knex was designed this way from the get-go, sure, such focus would be great and would position knex as a strictly "library for the library authors". However, doing this change now would be an abandonment of an existing huge userbase who rely on knex as a batteries-included solution.

**Comment by rijkvanzanten** [comment]
> Why would an average user want to manually pass generated SQL to a driver or use some additional wrapper on top of knex to execute the statement?

So it's connection library agnostic. In order to utilize `mysql2` instead of `mysql`, `knex` had to be updated to support it. Now that it does, the new kid on the block is `@mysql/xdevapi` which now creates more maintenance overhead in Knex to support.

> If knex was designed this way from the get-go, sure, such focus would be great and would position knex as a strictly "library for the library authors". However, doing this change now would be an abandonment of an existing huge userbase who rely on knex as a batteries-included solution.

I couldn't agree more with this. That being said, I'm noticing that a lot of these discussions end in a back and forth on what Knex is, was, and is supposed to be. I always saw/treated it as a batteries-included solution, but also noticed that (some) feature-requests and issues are closed with a (and i'm heavily paraphrasing here) "that's a difference in DB vendor, not a knex problem", which I also totally understand. It also seems like some of the unexpected behavior of knex is actually caused by these connection packages (like the mysql boolean example shared above), which further convolutes discussions around what knex does and should(n't) be doing.

In order to be a true batteries-included abstraction layer for all sql databases, we'd have to add pretty complex paths (like we've recently already done for SQLite schema manipulation) or what would be required to satisfy this OP issue. This ties directly into what @elhigu said above:

> Complex, non-trivial compatibility layers, should be done outside of knex as an external npm packages, where scope of those features and implementation will be well separated from the minimal core query building functionality of knex.

And (more importantly imo):

> Even with current really limited scope knex seems to be too big to be handled well with current contributors. With less bloat maybe we could even implement some important core features which would allow people implementing higher level packages to do better things

---

It feels like Knex currently sits in between. It's batteries-included _enough_ where one might assume it's a full database abstraction, only to run into IO/functionality differences in vendors that need to be worked around outside of Knex, or require pretty hacky workarounds to smoothen out in Knex itself, like https://github.com/knex/knex/pull/4322 or https://github.com/knex/knex/issues/4462.

I think it's important for the project to have a clear mission statement as to what Knex is supposed to be, so there's a clear "rule" of when to support what, and more importantly when _not_ to support things. If Knex is supposed to be a full batteries-included SQL database abstraction, I would expect methods to function the same and have the same IO across vendors. If it's purely a query builder, we can start using @elhigu's reasoning to more aggressively define the scope, and drastically reduce maintenance overhead of the main knex package.

> [...] that for its query builder instance it should return a single SQL + bindings that can be executed. Everything above that should be done in higher abstraction layers [...]

**Comment by kibertoad** [comment]
@rijkvanzanten I think that the best path forward for that would be to extract something like `knex-core` as a separate library and make `knex` depend on it for SQL building. Then it could be a non-breaking change for existing users. Not sure if there is anyone willing to contribute enough effort for that to happen, though.

**Comment by elhigu** [comment]
Would be nice if pool + connection handling and query building parts would be more separated. Usually people either wants the whole thing or just the query builder.

IMO having just the query builder as knex-query-builder package, that can be included without need for nodejs would be nice.  

Though complete knex would still have also nodejs dependent parts for crossdb compatible query and transaction execution.

**Comment by mercmobily** [comment]
Closing this in favour of #6254. I nearly kept this as the "main" issue. However, #6254  propose some practical solutions to the problem, and still references in bold characters. Please come over to #6254 to discuss!
